BAU: Fetch client config from SSM

Wynndow opened the pull request:
**Not to be merged before https://github.com/govuk-one-login/ipv-stubs-common-infra/pull/150**
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

Fetch client config from SSM

### Why did it change

This creates a new config service for fetching client config from the parameter store, rather than loading it in from a file.

This moves us towards how we manage config in the other components, and should make the set up for the CRI stub easier to understand.

The new ConfigService is currently only used for client config, but could be extended for all config in future work.

It's been implemented as a static class. This means it can just be called where it's needed and avoids having to inject it everywhere. This does make it harder to test, and a stub client has been created to allow some testing to happen.

Wynndow committed:
BAU: Fetch client config from SSM

This creates a new config service for fetching client config from the
parameter store, rather than loading it in from a file.

This moves us towards how we manage config in the other components, and
should make the set up for the CRI stub easier to understand.

The new ConfigService is currently only used for client config, but
could be extended for all config in future work.

It's been implemented as a static class. This means it can just be
called where it's needed and avoids having to inject it everywhere. This
does make it harder to test, and a stub client has been created to allow
some testing to happen.

Joe-Edwards-GDS commented on di-ipv-credential-issuer-stub/src/main/java/uk/gov/di/ipv/stub/cred/service/ConfigService.java:
```diff
@@ -0,0 +1,87 @@
+                || lastRefresh.isBefore(Instant.now().minusSeconds(getCacheDurationSeconds()))) {
+            LOGGER.info("Refreshing client configs");
+            getAllClientConfigs(null);
+            lastRefresh = Instant.now();
```
This is not thread-safe - do you think it needs to be?

In the most common cases it will probably just mean the config gets refreshed twice, but in the worst case it could corrupt the hash map.

You could avoid the latter case by making `getAllClientConfigs` build a fresh map, rather than mutating the original.

Wynndow replied:
Hadn't thought about threading - even though I don't think we need it right now it's probably a good idea to deal with it in case we do in future.

I've not done loads of multithreaded stuff, but I'm tempted to make the `refreshClientConfigsIfRequried` method `synchronized`. I think that should allow multiple threads to use the same `lastRefresh` timestamp - which would be advantageous, as well as protect the map? I'll update `getAllClientConfigs` to generate a new map rather than mutate too though.

Interested to understand how the hash map might get corrupted in the current implementation though?

Joe-Edwards-GDS replied:
Yeah, one neat thing about lambdas is that we usually don't have to worry about it (unless we deliberately write multithreaded code), but if this is running in a container and shared across requests, we probably do.

Making the method synchronised means that only one thread can enter the method at a time (and threads will just block until they can). In complex scenarios that can risk deadlocks etc. but that shouldn't be a worry here.

The issue previously is that if multiple threads attempt to read/update the map at the same time, you can get race conditions in the internal workings of the hashmap. I'm a bit rusty on all the details, but they usually have 'buckets' of hashes, and as you add/remove items, these can change in size. If a thread reads/updates during one of these operations, it could potentially lead to items getting lost, getting stale/unexpected values, etc.

Wynndow replied:
Cool, makes sense. Thanks Joe

Joe-Edwards-GDS commented on di-ipv-credential-issuer-stub/src/test/java/uk/gov/di/ipv/stub/cred/fixtures/TestFixtures.java:
```diff
@@ -29,9 +24,58 @@ private TestFixtures() {}
     public static final String DCMAW_VC =
             "eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NiJ9.eyJzdWIiOiJ1cm46dXVpZDo1ZDZkNjgzMy04NTEyLTRlMzctYjVlYS1iZTdkZTc3OTQ4ZGQiLCJhdWQiOiJodHRwczpcL1wvaWRlbnRpdHkuYnVpbGQuYWNjb3VudC5nb3YudWsiLCJuYmYiOjE2OTgxNDU2NjksImlzcyI6Imh0dHBzOlwvXC9kY21hdy1jcmkuc3R1YnMuYWNjb3VudC5nb3YudWsiLCJ2YyI6eyJ0eXBlIjpbIlZlcmlmaWFibGVDcmVkZW50aWFsIiwiSWRlbnRpdHlDaGVja0NyZWRlbnRpYWwiXSwiY3JlZGVudGlhbFN1YmplY3QiOnsibmFtZSI6W ...
+
+    public static final String CLIENT_CONFIG =
```
Could this just be a `ClientConfig` object, to avoid serialising/deserialising? I guess it'll be slightly clunkier in Java syntax :/

If not, I wonder if it'd be cleaner to load this as a JSON resource file, rather than embedding in Java.

Wynndow replied:
I think making it an object will be ok - will try it. I changed it from being b64 encoded to make it readable, but should have gone a step further!

Joe-Edwards-GDS reviewed (dismissed):
I'm not really a fan of the static singleton, but I can appreciate that it's much easier to implement this way(!)

Happy to leave it like this for now as it's no worse than the previous code, but left a couple of comments you might want to consider as well.

Wynndow committed:
BAU: Make ConfigService thread safe

This makes the refresh method synchronized. Whilst we're not using
multiple threads in the CRI stub at the moment, it would be good to
future proof it.

The syncronization should prevent different threads from calling the
method at the same time. This will allow them to use the same refresh
timestamp without issue, and should stop the configs map from being
updated multiple times.

Wynndow committed:
BAU: Make CLIENT_CONFIGS fixture a java object

To save the deserialization.

Wynndow committed:
BAU: Don't copy configs in Docker image

The configs file should no longer be required, and the GHA's have
stopped putting the file in place. We need to remove it from the
Dockerfile.
